libs/Facilitator/Facilitator.js

```diff
@@ -0,0 +1,387 @@
+    }
+
+    this.stakeAmount = new BN(amount);
+    if (this.stakeAmount.eqn(0)) {
+      throw new Error('Stake amount must not be zero.');
+    }
+
+    if (!web3.utils.isAddress(beneficiary)) {
+      throw new Error('Invalid beneficiary address.');
+    }
+
+    this.stakerAddress = staker;
+    this.facilitatorAddress = facilitator || staker;
+    this.hashLockObj = await this.getHashLock(unlockSecret);
+
+    // Get staker nonce.
+    this.stakerNonce = new BN(await this.getGatewayNonce(staker));
+
+    // Get Value token address.
+    const valueTokenAddress = await this.getValueToken();
+
+    // Get bounty amount.
+    this.bounty = new BN(await this.getBounty());
+
+    if (this.bounty.gtn(0)) {
+      // If bounty amount is not zero then get base token address.
+      const baseTokenAddress = await this.getBaseToken();
+
+      // If value and base token are same and if staker itself is facilitator,
+      // then we just have one approval. This will save some gas.
+      if (
+        valueTokenAddress === baseTokenAddress &&
+        this.stakerAddress === this.facilitatorAddress
+      ) {
+        const amountToApprove = this.stakeAmount.add(this.bounty);
+        await this.approveStakeAmount(
+          this.stakerAddress,
+          amountToApprove.toString(10),
+          gas
+        );
+      } else {
+        await this.approveBountyAmount(
+          this.facilitatorAddress,
+          this.bounty.toString(10)
+        );
+        await this.approveStakeAmount(
+          this.stakerAddress,
+          this.stakeAmount.toString(10),
+          gas
+        );
+      }
+    } else {
+      await this.approveStakeAmount(
+        this.stakerAddress,
+        this.stakeAmount.toString(10),
+        gas
+      );
+    }
+
+    const gatewayContract = this.contracts.Gateway(this.gatewayAddress);
+
+    const transactionOptions = {
+      from: this.facilitatorAddress,
+      to: this.gatewayAddress,
+      gas: gas
+    };
+
+    const tx = gatewayContract.methods.stake(
+      amount,
+      beneficiary,
+      gasPrice.toString(10),
+      gasLimit.toString(10),
+      this.stakerNonce.toString(10),
+      this.hashLockObj.hashLock
+    );
+
+    return this.sendTransaction(tx, transactionOptions);
+  }
+
+  /**
+   * @function getGatewayNonce
+   *
+   * Returns the gateway nonce for the given account address.
+   *
+   * @param {string} accountAddress Account address for which the nonce is to be fetched.
+   *
+   * @returns {Promise} Promise object represents the nonce of account address.
+   */
+  async getGatewayNonce(accountAddress) {
+    if (!web3.utils.isAddress(accountAddress)) {
+      throw new Error('Invalid account address.');
+    }
+
+    const gatewayContract = this.contracts.Gateway(this.gatewayAddress);
+
+    return gatewayContract.methods
+      .getNonce(accountAddress)
+      .call()
+      .then((nonce) => {
+        console.log(`\t - Gateway Nonce for ${accountAddress}:`, nonce);
```

Comment:
Can we please get rid of all the `console.log`s? They make it impossible to write an application with a clean output.
If we want to provide the additonal insight to library users that need it, we should be using a proper logging library like loglevel or winston.